### ISSUE TITLE
missing method for adding Variables at node in branching

### ISSUE THREAD

**Issue by networds** [SCIP-Interfaces/PySCIPOpt#289]
While implementing a Branch and Price procedure, I stumbled across the following problem:
When the pricer creates a new variable via `model.addVar` with the flag `priced = True` in a branching node that is not the root, the variable is somehow also available and added to the LP relaxation in other branching nodes where it is not compatible with the branching decisions there.

If I understand this correctly, I would need to create the variable with the `initial` flag of `SCIP` set to `False` as is done in the [VRP example from SCIP](https://scip.zib.de/doc/html/pricer__vrp_8cpp_source.php#l00249).
Is this correct?

Since PySCIPOpt currently only supports the method `model.addVar` which itself calls `SCIPcreateVarBasic` -- where there is no option to change the `initial` flag -- I have tried to add a function calling `SCIPcreateVar` myself.

Here is what I have:
```Python3
    def addVarAdvanced(self, name='', vtype='C', lb=0.0, ub=None, obj=0.0, pricedVar = False, initial = True, removable = False):
        """Create a new variable with the SCIP initial flag set to False. Default variable is non-negative and continuous.

        :param name: name of the variable, generic if empty (Default value = '')
        :param vtype: type of the variable (Default value = 'C')
        :param lb: lower bound of the variable, use None for -infinity (Default value = 0.0)
        :param ub: upper bound of the variable, use None for +infinity (Default value = None)
        :param obj: objective value of variable (Default value = 0.0)
        :param pricedVar: is the variable a pricing candidate? (Default value = False)
        :param initial: should var's column be present in the initial root LP?
        :param removable: is var's column removable from the LP (due to agiing or cleanup)?

        """

        # replace empty name with generic one
        if name == '':
            name = 'x'+str(SCIPgetNVars(self._scip)+1)

        cname = str_conversion(name)
        if lb is None:
            lb = -SCIPinfinity(self._scip)
        cdef SCIP_VAR* scip_var
        vtype = vtype.upper()
        if vtype in ['C', 'CONTINUOUS']:
            if ub is None:
                ub = SCIPinfinity(self._scip)
            PY_SCIP_CALL(SCIPcreateVar(self._scip, &scip_var, cname, lb, ub, obj, SCIP_VARTYPE_CONTINUOUS, initial, removable, NULL, NULL, NULL, NULL, NULL))
        elif vtype in ['B', 'BINARY']:
            if ub is None:
                ub = 1.0
            PY_SCIP_CALL(SCIPcreateVar(self._scip, &scip_var, cname, lb, ub, obj, SCIP_VARTYPE_BINARY, initial, removable, NULL, NULL, NULL, NULL, NULL))
        elif vtype in ['I', 'INTEGER']:
            if ub is None:
                ub = SCIPinfinity(self._scip)
            PY_SCIP_CALL(SCIPcreateVar(self._scip, &scip_var, cname, lb, ub, obj, SCIP_VARTYPE_INTEGER, initial, removable, NULL, NULL, NULL, NULL, NULL))
        else:
            raise Warning("unrecognized variable type")

        if pricedVar:
            PY_SCIP_CALL(SCIPaddPricedVar(self._scip, scip_var, 1.0))
        else:
            PY_SCIP_CALL(SCIPaddVar(self._scip, scip_var))

        pyVar = Variable.create(scip_var)

        #setting the variable data
        SCIPvarSetData(scip_var, <SCIP_VARDATA*>pyVar)
        PY_SCIP_CALL(SCIPreleaseVar(self._scip, &scip_var))
        return pyVar

```

I feel confident that this should do the trick in _scip.pyx_, however, I do not know what I have to add to _scip.pxd_ in order to make it work because of the `SCIP_DECL_VARDELORIG` etc. parameters.

So, there are two questions:

- is my understanding correct that creating a variable with initial=False does what I hope it does?
- if so, how can it be implemented in Cython/PySCIPOpt?

_edited: turn first bullet point into a meaningful sentence_

**Comment by fserra** [comment]
I don't think the `initial = False` will do the trick. Look here https://scip.zib.de/doc-6.0.2/html/FAQ.php#localvariables

**Comment by fserra** [comment]
The `SCIP_DECL_VARDELORIG` is a macro:
```
#define SCIP_DECL_VARDELORIG(x) SCIP_RETCODE x (SCIP* scip, SCIP_VAR* var, SCIP_VARDATA** vardata)
```
So one would need to write the macro out (expand it manually) for the argument.
Something like

```
SCIP_RETCODE (*vardelorig) (SCIP* scip, SCIP_VAR* var, SCIP_VARDATA** vardata)
```
We did that, for example, for `SCIPincludeReader`

But I don't think you really need that. Check the link that I posted above and, if you are lucky, the interface has everything you need already. Otherwise, you can try to extend it or we can do it together

**Comment by networds** [comment]
@fserra thank you for your helpful reply!

If I understand correctly, what I should do is write my custom constraint handler that, whenever the solver jumpes to a node in the branch and bound tree (let's say from node `A` to node `B`), checks the existing variables (specifically, the variables that were added to the LP after node `B` was created) for feasibility in the current node and set incompatible variables to `0` via the `consactive` callback?

In the original publication by Tobias Achterberg about SCIP something is mentioned that when switching from node `A` to node `B`, the solver backtracks the incrementally stored node data via a path in the branch and bound tree. Does this then mean that the `consactive` method is executed just for node `B` or for all nodes along the path from `A` to `B`?

Is there any way to avoid having to (slowly, since in Python) loop through all the variables for node `B`?

Thank you very much in advance!

PS: As a side question, I'd like to ask: if the `initial = False` is not meant for locally creating variables - what purpose does it serve?

**Comment by fserra** [comment]
Hi

From the documentation:
```
initial | should var's column be present in the initial root LP?
```
So, for example, if you have a lot of variables and not all of them appear in the first constraints that you are going to add to the LP at the beginning, then SCIP doesn't need to create a column for the variable.
You tell SCIP not to create the column by setting initial = False.

As for your other question, I think SCIP will only execute the active and deactive at node B and not along the path. 

You should actually have a look at the binpacking example in SCIP. You can find there a constraint handler that represents the branching decisions, which is what you want. A more elaborate example is in applications/coloring

Actually, in the branch `use-probdata-to-store-model` there is `test/test_coloring.py` which is a translation of the coloring application to python. This is not going to work, but you can still get some inspiration from there.

**Comment by networds** [comment]
Thanks a lot :)

Looking into the examples, I see that the `CONSACTIVE` callback is used to check whether the constraint should be repropapaged, and then in the propagation routine, the incompatible variables are set to `0`. Would it also be possible to to check the variables directly in the `CONSACTIVE` callback without using the propagation? Or would that not be executed during the correct phase/step in the algorithm?

**Comment by fserra** [comment]
You can find the answer in this email thread http://listserv.zib.de/pipermail/scip/2008-November/000077.html  
See also http://listserv.zib.de/pipermail/scip/2013-March/001357.html

From what I understood, you don't want to do that because of efficiency reasons and maybe is not even allowed.

**Comment by networds** [comment]
> As for your other question, I think SCIP will only execute the active and deactive at node B and not along the path.

this seems to be different; a print statement I have in the `consactive` method is indeed printed for each node along the path in the tree that is activated (that being all the nodes between the lowest common ancestor).

Nevertheless, I got the procedure to run now :) Thanks again for your help!
Let me summarize on what I did, for future reference:

- I added a dummy constraint handler
- In the `consactive` method, I check all the variables added since the node was created, whether they are compatible with the local brancing decisions
- If not, I set them to `0` using the `chgVarUbNode`
- Even though `consactive` is not only called for the current focus node but for all activated nodes (that contain my dummy constraint), I check, in `consactive`, whether the current node is the focus node
- therefore, I only need to change variable bounds in the current focus node
- since I only set variables to `0` in the focusnode, there is no reason to use the `consprop` method and scheduling the node for repropagation. Everything happens in the `consactive` method. Works like a charm.

I do not use the constraint handler to update my data structure for enforcing the branching rules. My data structure is a graph (using `graph_tool`). For each node, my branchrule updates and stores a `GraphView` on the parent node graph and attaches it to the node. Therefore, I always have the correct graph at hand in each node and do not need to make updates/changes when moving through the tree, as is done in some of the examples in the links or the `test/test_coloring.py` (in branch `use-probdata-to-store-model`).
